File: src/auth/store.ts
import fs from "fs/promises";
import path from "path";
import crypto from "crypto";
import type { StoredAuth, TokenSet } from "./types.js";

const ALGORITHM = "aes-256-gcm";

export class AuthStore {
  private dir: string;
  private encryptionKey: Buffer | null = null;

  constructor(baseDir: string) {
    this.dir = path.join(baseDir, ".auth");
  }

  async init(): Promise<void> {
    await fs.mkdir(this.dir, { recursive: true });

    // Derive encryption key from a machine-specific seed
    const keyFile = path.join(this.dir, ".key");
    try {
      const existing = await fs.readFile(keyFile);
      this.encryptionKey = existing;
    } catch {
      const key = crypto.randomBytes(32);
      await fs.writeFile(keyFile, key, { mode: 0o600 });
      this.encryptionKey = key;
    }
  }

  async save(provider: string, tokens: TokenSet): Promise<void> {
    const auth: StoredAuth = {
      provider,
      tokens,
      created: new Date().toISOString(),
      updated: new Date().toISOString(),
    };

    const encrypted = this.encrypt(JSON.stringify(auth));
    const filePath = path.join(this.dir, `${provider}.enc`);
    await fs.writeFile(filePath, encrypted, "utf-8");
  }

  async load(provider: string): Promise<StoredAuth | null> {
    const filePath = path.join(this.dir, `${provider}.enc`);
    try {
      const encrypted = await fs.readFile(filePath, "utf-8");
      const decrypted = this.decrypt(encrypted);
      return JSON.parse(decrypted);
    } catch {
      return null;
    }
  }

  async delete(provider: string): Promise<boolean> {
    const filePath = path.join(this.dir, `${provider}.enc`);
    try {
      await fs.unlink(filePath);
      return true;
    } catch {
      return false;
    }
  }

  async list(): Promise<string[]> {
    try {
      const files = await fs.readdir(this.dir);
      return files
        .filter((f) => f.endsWith(".enc"))
        .map((f) => f.replace(".enc", ""));
    } catch {
      return [];
    }
  }

  private encrypt(text: string): string {
    if (!this.encryptionKey) throw new Error("AuthStore not initialized");
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(ALGORITHM, this.encryptionKey, iv);
    let encrypted = cipher.update(text, "utf8", "hex");
    encrypted += cipher.final("hex");
    const tag = cipher.getAuthTag();
    return `${iv.toString("hex")}:${tag.toString("hex")}:${encrypted}`;
  }

  private decrypt(text: string): string {
    if (!this.encryptionKey) throw new Error("AuthStore not initialized");
    const [ivHex, tagHex, encrypted] = text.split(":");
    const iv = Buffer.from(ivHex, "hex");
    const tag = Buffer.from(tagHex, "hex");
    const decipher = crypto.createDecipheriv(ALGORITHM, this.encryptionKey, iv);
    decipher.setAuthTag(tag);
    let decrypted = decipher.update(encrypted, "hex", "utf8");
    decrypted += decipher.final("utf8");
    return decrypted;
  }
}

File: synesis/auth/store.py
from __future__ import annotations

import json
import os
from cryptography.fernet import Fernet
from pathlib import Path


class AuthStore:
    def __init__(self, base_dir: str | Path):
        self.dir = Path(base_dir) / ".auth"
        self._fernet: Fernet | None = None

    def init(self) -> None:
        self.dir.mkdir(parents=True, exist_ok=True)
        os.chmod(self.dir, 0o700)
        key_file = self.dir / ".key"

        if key_file.exists():
            key = key_file.read_bytes()
        else:
            key = Fernet.generate_key()
            key_file.write_bytes(key)
            os.chmod(key_file, 0o600)

        self._fernet = Fernet(key)

    def save(self, provider: str, tokens: dict) -> None:
        if not self._fernet:
            raise RuntimeError("AuthStore not initialized")

        data = json.dumps(tokens).encode()
        encrypted = self._fernet.encrypt(data)
        (self.dir / f"{provider}.enc").write_bytes(encrypted)

    def load(self, provider: str) -> dict | None:
        if not self._fernet:
            raise RuntimeError("AuthStore not initialized")

        path = self.dir / f"{provider}.enc"
        if not path.exists():
            return None

        try:
            encrypted = path.read_bytes()
            decrypted = self._fernet.decrypt(encrypted)
            return json.loads(decrypted)
        except Exception:
            return None

    def delete(self, provider: str) -> bool:
        path = self.dir / f"{provider}.enc"
        if path.exists():
            path.unlink()
            return True
        return False

    def list(self) -> list[str]:
        if not self.dir.exists():
            return []
        return [f.stem for f in self.dir.glob("*.enc")]
